### ISSUE TITLE
communicate multiple times

### ISSUE THREAD

**Issue by ShaneEverittM** [hniksic/rust-subprocess#46]
I believe under the current implementation there is no way to call communicate multiple times to a subprocess that terminates on EOF. It would be nice if there was some sort of mechanism by which I could write to a subprocess without sending eof, such that I could read its response and send more based on the response. I apologize if this is already possible, or if my lack of understanding of the implementation is preventing me from seeing if this is impossible.

**Comment by hniksic** [comment]
That is possible by directly reading and writing to the subprocess's `stdin`, `stdout` and `stderr`, all exposed by the `Popen` struct. `communicate()` is just a convenience function that uses the public API to implement a simple communication with the properties that:

* the data written to subprocess and the data read from the subprocess fits in memory
* all the data from subprocess is read (possibly limited by size and timeout)
* the data is written and read in parallel to avoid deadlocks

More sophisticated communication, such as exchanging multiple messages, can be implemented by using the IO handles directly, as noted above. Note that such communication likely requires the process and the subprocess to agree on some protocol for message exchange, even if a simple one (e.g. exchanging self-delimited messages). That is not hard to implement and does not need a generic framework to support; if you can provide more detail about your use case, I might be able to offer more concrete advice.

**Comment by ShaneEverittM** [comment]
Thanks so much for the fast response. My use case is pretty specific. I've written a couple of applications that have a shell-like interface, with various commands they accept and output based on commands, you get the idea. I want to show of these projects on a website I'm building, and the way I thought to accomplish this was by basically forwarding all input from a frontend construct that looks like a terminal to the backend, and having it spawn and keep alive the project I want to show off, giving it the command from the frontend, and sending back the response. The annoying thing is I don't really want to go back and edit those projects to create a communication protocol, but I will if necessary.

**Comment by hniksic** [comment]
That use case should work with `subprocess`, you just need to send commands to `p.stdin` and read output from `p.stdout`, `p` being the subprocess. Your "protocol" consists of lines being sent and received.

The deadlock that `communicate()` is designed to prevent comes from slightly more involved use cases where a subprocess responds to what you're writing before you're done writing. With a line-based interface based on shell-like commands, you can just read and write in a natural fashion.

**Comment by ChrisCA** [comment]
I coincidentally came along this topic while I was writing a quick and dirty wrapper for the bluetoothctl command line tool.
Just wanted share the snippet which might not be most beautiful but should be an example for what you need @ShaneEverittM 

```
        let p = Popen::create(
            &["bluetoothctl"],
            PopenConfig {
                stdout: Redirection::Pipe,
                stdin: Redirection::Pipe,
                ..Default::default()
            },
        )
        .expect("couldn't spawn child command");

        let mut lines = BufReader::new(p.stdout.as_ref().unwrap()).lines();
        std::thread::sleep(Duration::from_millis(500)); // give bluetoothctl some time to start until issue the scan on command
        let _ = p.stdin.as_ref().unwrap().write_all("scan on\n".as_bytes());
```

**Comment by hniksic** [comment]
@ChrisCA That's the what I had in mind, thanks.

Side note: I think the sleep is unnecessary, you can simply write the data and wait for `bluetoothctl` to process it - the data won't (or shouldn't) be forgotten if you write it a bit sooner. Also, `let _ = <...>.write_all(...)` will silently ignore the errors in writing, which doesn't sound like a good idea. If panic is undesirable and the program needs to continue despite the error, it can still be reported using something like `.map_err(|e| eprintln!("error writing: {}", e))`.

**Comment by ShaneEverittM** [comment]
Wow, thanks so much for the helpful info! Works perfectly. Honestly you made me remember to explicitly send a "\n", so I feel foolish.

**Comment by martonlanga** [comment]
Hey @hniksic, I'm trying to create something that works kind of like a shell using subprocess. A user enters a command in a webview and this executes the command in a subprocess inside a new thread. I'm trying to communicate with them using channels from the main thread and send messages from there to the subprocesses.
I've managed to make stdout/err work like this but I'm having trouble with stdin.

I've tried a lot of things, here's my current attempt:

```rust
let mut p = Exec::cmd(command)
    .args(&args)
    .cwd(current_dir)
    .stdin(Redirection::Pipe)
    .stdout(Redirection::Pipe)
    .stderr(Redirection::Pipe)
    .popen()?;

let mut stdin = vec![];

while p.poll().is_none() {
    // trying to write stdin (this is not working)
    p.stdin
        .as_ref()
        .unwrap()
        .write_all(&stdin)
        .expect("Failed to write stdin");

    let mut reader = BufReader::new(p.stdout.as_ref().unwrap());
    let mut chunk = [0u8; 64];
    loop {
        let len = reader.read(&mut chunk)?;
        if len == 0 {
            break;
        }
        let chunk = &chunk[..len];

        // send stdout
        let event_result = tauri::event::emit(
            webview,
            "event",
            Some(Payload {
                id: id.clone(),
                chunk,
            }),
        );
        if let Err(err) = event_result {
            println!("Failed to send chunk: {}", err);
        } else {
            println!("Sent chunk");
        };
    }

    // receive stdin message
    let received = receiver.recv();
    if let Ok(message) = received {
        if message == "Exit" {
            println!("Exiting");
            break;
        }

        println!("Message received: {}", message);
        stdin = message.as_bytes().to_owned();
    } else {
        println!("No message or error");
    }
}
```

Please let me know if it's unclear what I'm trying to do. Any help would be appreciated, and thanks for your time!

**Comment by hniksic** [comment]
@martonlanga Thanks for the report, but could you please open a separate issue for this? It looks quite unrelated to this issue, which is about the `communicate()` method.
